src/menus/PluginMenus.cpp

```diff
@@ -62,6 +64,8 @@ void DoManageRealtimeEffectsSidePanel(AudacityProject &project)
 
 }
 
+extern IntSetting SelectionToolbarMode;
```

Comment:
You removed the repetition of the path literal, which is good.

But you introduced an `extern` declaration, which I don't like.

It lets you "cheat" without properly declaring this setting in a header file, and including that file, so that the tool that produces the "jungle" graph doesn't reflect the complete, real dependencies.

Please declare the setting in SelectionBar.h.

That does not add dependency on SelectionBar -- the dependency was already there because of this extern; now it is made more obvious.

And we may not like adding dependencies on the toolbars, but from a src/menus file it is  not bad.  These files are already tops of the dependency graph -- note that they are .cpp files without corresponding .h and rely on registration to add their functionality.

Reply:
Please do this

Reply:
I think this is a missing feature of the Setting classes. They should reset when the configuration resets. The need to manually reset preferences wrapped in Setting violates Open/Closed - every time you add such preference you end up updating PluginMenus. Ideally `CachingSettingBase` should listen to reset preferences events and call `Invalidate`

Reply:
I have formed a different opinion:  I blame how ShuttleGui presently handles Settings objects.  It bypasses them and directly uses gPrefs->Write() directly on the path.  That happens in ShuttelPrefs.cpp.

Instead, in all places where the Tie... functions are called with Settings object, these should be rewritten always as named Settings objects, cleaning up many preference paths that are not yet Settings, and passed as non-const references.  Then, rewrite ShuttleGui to use the interface of Settings only; and then Invalidate() will become unnecessary.

The waiting work to do this cleanup is already in my OTHER monster branch attempting a revised ShuttleGui and would be the first stage of that ambitious project.

Reply:
I don't  think we are talking about the same thing.

I am talking about `ResetPreferences` not actually resetting any of the `CachingSettingBase`, I don't see how changing ShuttleGui will help to solve this problem. For this reason `OnResetConfig(const CommandContext &context)` growth every time new `CachingSettingBase` setting is created (if one cares about resetting of this value). BTW this is an example, where everything is scattered between translation units, but Open/Closed is violated. 

On top of that, it creates inconsistent behavior: if specific `CachingSettingBase` is not manually added to `OnResetConfig` there is a different state between gPrefs and `CachingSettingBase`. If that setting is not modified, it will be reset on the next launch, so it won't be kept.

At very list we need a global message that preferences were reset.

Reply:
I see it now, and agree.

I think Invalidate() should be made unnecessary.  I did not remember this other necessary part of that work.

Generally, naked `gPrefs` must be discouraged in new code and fixed in a big cleanup.

Open/Closed, correctly done, allows things to scatter.  I never claimed that conversely, if things are scattered, therefore Open/Closed is realized.

In fact I never did claim Settings objects in particular as an example of application of Open/Closed as I understand it.  But I want them for the other reason that magic literals (here not numbers but string literals) should be avoided.

Reply:
What bugs can be demonstrated?  In fact this should not be difficult to fix now.  Make every Settings object an observer of a singleton Publisher that fires after `gPrefs->DeleteAll().`

Reply:
> In fact this should not be difficult to fix now

Not difficult,  but not for now. I expect extensive testing. 

> What bugs can be demonstrated?

1.  Remove the line we are discussing
2.  Change the selection mode
3. Reset the configs using menu
4. Observe selection mode has not changed
5. Relaunch Audacity
6. Observe selection mode changed to deafult.

I expect that there are other similar problems, or problems with the stickiness which are not so easy to observe.